File: minigit-5/ts/src/commands/branch.ts
// branch command - List, create, or delete branches

import { findRepoRoot } from '../utils';
import { listBranches, createBranch, deleteBranch, renameBranch, getCurrentBranch, getHeadCommit, resolveRef, resolveRevision } from '../refs';
import { getCommit } from '../objects';

export function branch(args: string[]): number {
  const repoRoot = findRepoRoot();
  if (!repoRoot) {
    console.error('fatal: not a minigit repository');
    return 1;
  }

  // Parse arguments
  let deleteMode = false;
  let forceDelete = false;
  let renameMode = false;
  let verbose = false;
  const positionalArgs: string[] = [];

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg === '-d') {
      deleteMode = true;
    } else if (arg === '-D') {
      deleteMode = true;
      forceDelete = true;
    } else if (arg === '-m') {
      renameMode = true;
    } else if (arg === '-v' || arg === '--verbose') {
      verbose = true;
    } else if (!arg.startsWith('-')) {
      positionalArgs.push(arg);
    }
  }

  // Handle delete mode
  if (deleteMode) {
    if (positionalArgs.length === 0) {
      console.error('fatal: branch name required');
      return 1;
    }

    const branchName = positionalArgs[0];
    const currentBranch = getCurrentBranch(repoRoot);

    if (branchName === currentBranch) {
      console.error(`error: Cannot delete branch '${branchName}' checked out`);
      return 1;
    }

    // Check if branch exists
    const branchSha = resolveRef(`refs/heads/${branchName}`, repoRoot);
    if (!branchSha) {
      console.error(`error: branch '${branchName}' not found.`);
      return 1;
    }

    // For -d (not -D), check if branch is merged into HEAD
    if (!forceDelete) {
      const headSha = getHeadCommit(repoRoot);
      if (headSha && !isAncestor(branchSha, headSha, repoRoot)) {
        console.error(`error: The branch '${branchName}' is not fully merged.`);
        console.error(`If you are sure you want to delete it, run 'minigit branch -D ${branchName}'.`);
        return 1;
      }
    }

    try {
      deleteBranch(branchName, repoRoot);
      console.log(`Deleted branch ${branchName} (was ${branchSha.slice(0, 7)}).`);
      return 0;
    } catch (e: any) {
      console.error(`error: ${e.message}`);
      return 1;
    }
  }

  // Handle rename mode
  if (renameMode) {
    if (positionalArgs.length < 2) {
      console.error('fatal: branch rename requires old and new names');
      return 1;
    }

    const oldName = positionalArgs[0];
    const newName = positionalArgs[1];

    try {
      renameBranch(oldName, newName, repoRoot);
      console.log(`Branch '${oldName}' renamed to '${newName}'.`);
      return 0;
    } catch (e: any) {
      console.error(`error: ${e.message}`);
      return 1;
    }
  }

  // Handle create mode
  if (positionalArgs.length > 0) {
    const branchName = positionalArgs[0];

    // Validate branch name
    if (branchName.startsWith('-') || branchName.includes(' ') || branchName.startsWith('.') || branchName.includes('..')) {
      console.error(`fatal: '${branchName}' is not a valid branch name.`);
      return 1;
    }

    let startPoint: string;
    if (positionalArgs.length > 1) {
      try {
        startPoint = resolveRevision(positionalArgs[1], repoRoot);
      } catch (e: any) {
        console.error(`fatal: ${e.message}`);
        return 1;
      }
    } else {
      const headSha = getHeadCommit(repoRoot);
      if (!headSha) {
        console.error('fatal: Not a valid object name: HEAD');
        return 1;
      }
      startPoint = headSha;
    }

    try {
      createBranch(branchName, startPoint, repoRoot);
      return 0;
    } catch (e: any) {
      console.error(`fatal: ${e.message}`);
      return 1;
    }
  }

  // List mode
  const branches = listBranches(repoRoot);
  const currentBranch = getCurrentBranch(repoRoot);

  branches.sort();

  for (const branch of branches) {
    const isCurrent = branch === currentBranch;
    const prefix = isCurrent ? '* ' : '  ';

    if (verbose) {
      const sha = resolveRef(`refs/heads/${branch}`, repoRoot);
      if (sha) {
        try {
          const commit = getCommit(sha, repoRoot);
          const shortSha = sha.slice(0, 7);
          const firstLine = commit.message.split('\n')[0];
          console.log(`${prefix}${branch} ${shortSha} ${firstLine}`);
        } catch {
          console.log(`${prefix}${branch}`);
        }
      } else {
        console.log(`${prefix}${branch}`);
      }
    } else {
      console.log(`${prefix}${branch}`);
    }
  }

  return 0;
}

function isAncestor(ancestor: string, descendant: string, repoRoot: string): boolean {
  // BFS from descendant to find ancestor
  const visited = new Set<string>();
  const queue = [descendant];

  while (queue.length > 0) {
    const sha = queue.shift()!;
    if (visited.has(sha)) continue;
    visited.add(sha);

    if (sha === ancestor) {
      return true;
    }

    try {
      const commit = getCommit(sha, repoRoot);
      queue.push(...commit.parents);
    } catch {
      // Invalid commit
    }
  }

  return false;
}

File: minigit-5/py/commands/branch.py
"""branch command - List, create, or delete branches."""

import sys
import re
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent))

from utils import find_repo_root
from objects import Commit
from refs import (
    list_branches, get_current_branch, read_ref, write_ref, delete_ref,
    resolve_revision
)


def run(args: list[str]) -> int:
    """Manage branches."""
    repo_root = find_repo_root()
    if repo_root is None:
        print("fatal: not a minigit repository", file=sys.stderr)
        return 1

    # Parse arguments
    delete = False
    force_delete = False
    rename = False
    verbose = False
    branch_name = None
    start_point = None

    i = 0
    while i < len(args):
        arg = args[i]
        if arg == '-d':
            delete = True
        elif arg == '-D':
            force_delete = True
        elif arg == '-m':
            rename = True
        elif arg == '-v':
            verbose = True
        elif not arg.startswith('-'):
            if branch_name is None:
                branch_name = arg
            else:
                start_point = arg
        i += 1

    if delete or force_delete:
        if not branch_name:
            print("fatal: branch name required", file=sys.stderr)
            return 1
        return delete_branch(repo_root, branch_name, force_delete)

    if rename:
        if not branch_name or not start_point:
            print("fatal: rename requires old and new names", file=sys.stderr)
            return 1
        return rename_branch(repo_root, branch_name, start_point)

    if branch_name:
        return create_branch(repo_root, branch_name, start_point)

    return list_branches_cmd(repo_root, verbose)


def list_branches_cmd(repo_root: Path, verbose: bool) -> int:
    """List all branches."""
    branches = list_branches(repo_root)
    current = get_current_branch(repo_root)

    for branch in branches:
        if branch == current:
            prefix = '* '
        else:
            prefix = '  '

        if verbose:
            sha = read_ref(repo_root, f'refs/heads/{branch}')
            if sha:
                try:
                    commit = Commit.read(sha, repo_root)
                    message = commit.message.split('\n')[0]
                    print(f"{prefix}{branch} {sha[:7]} {message}")
                except:
                    print(f"{prefix}{branch} {sha[:7]}")
            else:
                print(f"{prefix}{branch}")
        else:
            print(f"{prefix}{branch}")

    return 0


def create_branch(repo_root: Path, name: str, start_point: str = None) -> int:
    """Create a new branch."""
    # Validate branch name
    if not is_valid_branch_name(name):
        print(f"fatal: invalid branch name: {name}", file=sys.stderr)
        return 1

    # Check if branch already exists
    if read_ref(repo_root, f'refs/heads/{name}'):
        print(f"fatal: a branch named '{name}' already exists", file=sys.stderr)
        return 1

    # Get starting point SHA
    if start_point:
        sha = resolve_revision(repo_root, start_point)
        if sha is None:
            print(f"fatal: not a valid object name: '{start_point}'", file=sys.stderr)
            return 1
    else:
        from refs import resolve_head
        sha = resolve_head(repo_root)
        if sha is None:
            print("fatal: not a valid object name: 'HEAD'", file=sys.stderr)
            return 1

    # Create the branch
    write_ref(repo_root, f'refs/heads/{name}', sha)
    return 0


def delete_branch(repo_root: Path, name: str, force: bool) -> int:
    """Delete a branch."""
    current = get_current_branch(repo_root)

    if name == current:
        print(f"error: cannot delete branch '{name}' used by worktree", file=sys.stderr)
        return 1

    ref_path = f'refs/heads/{name}'
    if not read_ref(repo_root, ref_path):
        print(f"error: branch '{name}' not found", file=sys.stderr)
        return 1

    if not force:
        # Check if branch is merged
        from refs import resolve_head
        head_sha = resolve_head(repo_root)
        branch_sha = read_ref(repo_root, ref_path)

        if head_sha and branch_sha and not is_ancestor(repo_root, branch_sha, head_sha):
            print(f"error: branch '{name}' is not fully merged", file=sys.stderr)
            print(f"If you are sure you want to delete it, run 'branch -D {name}'", file=sys.stderr)
            return 1

    delete_ref(repo_root, ref_path)
    print(f"Deleted branch {name}")
    return 0


def rename_branch(repo_root: Path, old_name: str, new_name: str) -> int:
    """Rename a branch."""
    if not is_valid_branch_name(new_name):
        print(f"fatal: invalid branch name: {new_name}", file=sys.stderr)
        return 1

    old_ref = f'refs/heads/{old_name}'
    new_ref = f'refs/heads/{new_name}'

    sha = read_ref(repo_root, old_ref)
    if not sha:
        print(f"error: branch '{old_name}' not found", file=sys.stderr)
        return 1

    if read_ref(repo_root, new_ref):
        print(f"fatal: a branch named '{new_name}' already exists", file=sys.stderr)
        return 1

    # Create new ref and delete old
    write_ref(repo_root, new_ref, sha)
    delete_ref(repo_root, old_ref)

    # Update HEAD if it was pointing to old branch
    current = get_current_branch(repo_root)
    if current == old_name:
        from refs import write_head
        write_head(repo_root, f'ref: refs/heads/{new_name}')

    return 0


def is_valid_branch_name(name: str) -> bool:
    """Check if branch name is valid."""
    if not name:
        return False
    if name.startswith('-'):
        return False
    if ' ' in name:
        return False
    if '..' in name:
        return False
    if name.startswith('.'):
        return False
    if name.endswith('/'):
        return False
    if name.endswith('.lock'):
        return False
    if '@{' in name:
        return False
    # Check for invalid characters
    invalid_chars = ['~', '^', ':', '\\', '*', '?', '[']
    for char in invalid_chars:
        if char in name:
            return False
    return True


def is_ancestor(repo_root: Path, commit1: str, commit2: str) -> bool:
    """Check if commit1 is an ancestor of commit2."""
    visited = set()
    queue = [commit2]

    while queue:
        sha = queue.pop(0)
        if sha in visited:
            continue
        visited.add(sha)

        if sha == commit1:
            return True

        try:
            commit = Commit.read(sha, repo_root)
            queue.extend(commit.parents)
        except:
            pass

    return False
